### DISCUSSION TITLE
Filter out results from consult-grep/ripgrep etc with !filter-string

### DISCUSSION THREAD

**Discussion by mattiasdrp** [minad/consult#1119] [Q&A] [answered]
Hi minad,

I'm trying to filter out results from my search but it looks like the way I do it is not working well:

![image](https://github.com/user-attachments/assets/be60029f-eb32-4070-b9cc-55138c047c05)

If I try to filter out lines with `pokemacs`:

![image](https://github.com/user-attachments/assets/42d83f4c-9c7d-41f1-9fac-2be6155e37ad)

And it works as expected when I try to keep them:

![image](https://github.com/user-attachments/assets/6a54a30e-a3ee-4f77-aeb6-70415d2598d0)

Am I doing something wrong or can this only be used to filter lines containing the second part of the input string?

**Comment by minad** [comment]
Please try to reproduce the issue with a minimal setup. See https://github.com/minad/consult?tab=readme-ov-file#bug-reports.

**Comment by mattiasdrp** [comment]
With the following init.el file:

```
(defvar elpaca-installer-version 0.7)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil :depth 1
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca--activate-package)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (< emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                 ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                 ,@(when-let ((depth (plist-get order :depth)))
                                                     (list (format "--depth=%d" depth) "--no-single-branch"))
                                                 ,(plist-get order :repo) ,repo))))
                 ((zerop (call-process "git" nil buffer t "checkout"
                                       (or (plist-get order :ref) "--"))))
                 (emacs (concat invocation-directory invocation-name))
                 ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                       "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                 ((require 'elpaca))
                 ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (load "./elpaca-autoloads")))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))

(elpaca elpaca-use-package
  ;; Enable :ensure use-package keyword.
  (elpaca-use-package-mode)
  ;; Assume :ensure t unless otherwise specified.
  (setq elpaca-use-package-by-default t))
(elpaca-wait)

(use-package vertico
  :ensure (vertico :files (:defaults "extensions/*"))
  :init (vertico-mode)
  :config (message "`vertico' loaded"))

(use-package consult
  :ensure t
  :config (message "`vertico' loaded"))

;;;; Footer

;; End:
(provide 'init)

```

`M-x consult-ripgrep RET (defun`:

![image](https://github.com/user-attachments/assets/24a82ced-bd07-4b69-b623-7087546e2f77)

`M-x consult-ripgrep RET (defun#!elpaca`:

![image](https://github.com/user-attachments/assets/13288bd8-1de7-4ae6-9c3b-b9a69dc2d12a)

**Reply by minad** [reply]
Try `vertico-group-format=nil` and then you will see the reason for the behavior.

**Reply by mattiasdrp** [reply]
I think this answer was for the next comment about consult filtering filenames. Thanks for the tip ;-)

**Comment by mattiasdrp** [comment]
If my cursor is just after the second `#` it displays the same result as if I wrote a search without the filter-string part:

![image](https://github.com/user-attachments/assets/476713cd-b597-4438-bc2a-3f7cf7139ead)

(Sorry, this is only true if I don't use

```
(setq completion-styles '(substring basic))
```

If I use it:

It filters according to the filename so nothing is filtered out when using `elpaca`:

![image](https://github.com/user-attachments/assets/bb1082f6-405d-4549-ba14-f986966348b0)

`vertico` is a good candidate:

![image](https://github.com/user-attachments/assets/c4f4242e-8670-4817-986b-358d0a2a84db)

But `!vertico` doesn't work:

![image](https://github.com/user-attachments/assets/6f55fef8-6cf4-44e9-b2de-5fd14192fe24)

**Comment by mattiasdrp** [comment]
I don't understand how your minimal setup is supposed to work so I used a custom one based on lsp-plain.el from lsp-mode:

```
(require 'package)
(package-initialize)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("gnu" . "https://elpa.gnu.org/packages/"))
      package-user-dir (make-temp-file "vertico-tmp-elpa")
      custom-file (expand-file-name "custom.el" package-user-dir))
(delete-file package-user-dir)
(let* ((pkg-list '(vertico consult)))
  (package-initialize)
  (package-refresh-contents)
  (mapc (lambda (pkg)
          (unless (package-installed-p pkg)
            (package-install pkg))
          (require pkg))
        pkg-list)
  (vertico-mode)
  (setq completion-styles '(substring basic)))
```

And it gives me the same result

**Comment by oantolin** [comment]
I see in these setups that the completion styles used are substring and basic, neither of which supports `!` to filter out. Try again using a completion style.like orderless (install it first) which does.

**Reply by mattiasdrp** [reply]
Ok, thanks :-) I have this completion style in my init.el file but I still have this issue so I need to find what's causing it

**Comment by mattiasdrp** [comment]
Ok, I'm so stupid. I used `(substring orderless basic)` as completion style instead of `(orderless basic)`

Sorry for the noise

**Comment by mattiasdrp** [comment, marked as answer]
Ok so the reason why it didn't work in my setup is that I had these lines in my lsp-mode config:

```
   (defun minad/orderless-dispatch-prefixes-first (_pattern index _total)
     (and (eq index 0) 'orderless-prefixes))

   (add-hook 'orderless-style-dispatchers #'minad/orderless-dispatch-prefixes-first)
```

I don't remember exactly what they did but commenting them brought back the behaviour of orderless for string filtering. I tried to find in the different wikis (consult, corfu, vertico) if I could find these lines but there have been either removed or they're in a different place.
